### PR TITLE
test: space roles

### PR DESCRIPTION
**Pull request by xcoulon**
E2E tests for https://github.com/codeready-toolchain/member-operator/pull/327

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: test/e2e/space_test.go [hunk 1]
```diff
@@ -121,18 +120,138 @@ func TestCreateSpace(t *testing.T) {
...
+		wait.UntilSpaceHasConditions(Provisioned()),
+		wait.UntilSpaceHasStatusTargetCluster(memberAwait.ClusterName),
+	)
+	require.NoError(t, err)
+	// wait until NSTemplateSet has been created and Space is in `Ready` status
+	nsTmplSet, err := memberAwait.WaitForNSTmplSet(s.Name, wait.UntilNSTemplateSetHasConditions(Provisioned()))
+	require.NoError(t, err)
+	tierChecks, err := tiers.NewChecks("appstudio")
+	require.NoError(t, err)
+	tiers.VerifyGivenNsTemplateSet(t, memberAwait, nsTmplSet, tierChecks, tierChecks, tierChecks.GetExpectedTemplateRefs(hostAwait))
```
Comment (MatousJobanek) [inline test/e2e/space_test.go:149]: you could use the function `VerifyResourcesProvisionedForSpaceWithTier`
```suggestion
	// wait until NSTemplateSet has been created and Space is in `Ready` status
	VerifyResourcesProvisionedForSpaceWithTier(t, awaitilities, memberAwait, s.Name, "appstudio")
```

File: test/e2e/space_test.go [hunk 2]
```diff
@@ -121,18 +120,138 @@ func TestCreateSpace(t *testing.T) {
...
+
+		// when
+		err = memberAwait.Client.Update(context.TODO(), nsTmplSet)
+
+		// then
+		require.NoError(t, err)
+		_, err = memberAwait.WaitForNSTmplSet(nsTmplSet.Name, wait.UntilNSTemplateSetHasConditions(Provisioned()))
+		require.NoError(t, err)
+		ns, err := memberAwait.WaitForNamespace(s.Name, nsTmplSet.Spec.Namespaces[0].TemplateRef, "appstudio")
+		require.NoError(t, err)
```
Comment (MatousJobanek) [inline test/e2e/space_test.go:171]: could we also check the content of the annotation?

File: test/e2e/space_test.go [hunk 3]
```diff
@@ -121,18 +120,138 @@ func TestCreateSpace(t *testing.T) {
...
+		nsTmplSet.Spec.SpaceRoles = []toolchainv1alpha1.NSTemplateSetSpaceRole{
+			{
+				TemplateRef: adminTierTmpl.Name,
+				Usernames:   []string{"user1", "user2"},
+			},
+		}
+
+		// when
+		err = memberAwait.Client.Update(context.TODO(), nsTmplSet)
```
Comment (MatousJobanek) [inline test/e2e/space_test.go:164]: btw, you can again hit the conflict error when updating the NSTemplateSet

File: testsupport/wait/member.go [hunk 4]
```diff
@@ -432,6 +432,21 @@ func (a *MemberAwaitility) WaitForRoleBinding(namespace *corev1.Namespace, name
 	return roleBinding, err
 }
 
+// WaitUntilRoleBindingDeleted waits until a Role with the given name does not exist anymore in the given namespace
```
Comment (MatousJobanek) [inline testsupport/wait/member.go:435]:
```suggestion
// WaitUntilRoleBindingDeleted waits until a RoleBinding with the given name does not exist anymore in the given namespace
```

**Review by MatousJobanek — COMMENTED** [review]
Cool :+1: but please be careful with the updates of the NSTemplateset

File: test/e2e/space_test.go (hunk 3 above)
Comment (xcoulon) [reply test/e2e/space_test.go]: yes, I saw that too. I've added an `UpdateNSTemplateSet` method in `wait.MemberAwaitility` to fix that. See 89831dd

File: testsupport/wait/member.go [hunk 5]
```diff
@@ -344,8 +345,53 @@ func (a *MemberAwaitility) WaitUntilNSTemplateSetDeleted(name string) error {
...
+			if !found {
+				return false
+			}
+			expectedLastAppliedSpaceRoles, _ := json.Marshal(expected) // assume that encoding always works
+			return string(expectedLastAppliedSpaceRoles) == lastAppliedSpaceRoles
+		},
+		Diff: func(actual *corev1.Namespace) string {
+			return "expected namespace to be match annotation"
+		},
+	}
```
Comment (MatousJobanek) [inline testsupport/wait/member.go:381]: wouldn't make this the diff better?
```suggestion
	expectedLastAppliedSpaceRoles, _ := json.Marshal(expected) // assume that encoding always works
	return NamespaceWaitCriterion{
		Match: func(actual *corev1.Namespace) bool {
			lastAppliedSpaceRoles, found := actual.Annotations[toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey]
			if !found {
				return false
			}

			return string(expectedLastAppliedSpaceRoles) == lastAppliedSpaceRoles
		},
		Diff: func(actual *corev1.Namespace) string {
			return fmt.Sprintf("expected namespace to be match annotation,\nExpected: %s\nActual annotations:%v", expectedLastAppliedSpaceRoles, actual.Annotations)
		},
	}
```

File: go.mod [hunk 6]
```diff
@@ -35,4 +35,6 @@ require (
 	sigs.k8s.io/controller-runtime v0.8.3
 )
 
+replace github.com/codeready-toolchain/api => github.com/xcoulon/api v0.0.0-20220120072908-2aafc9d64340
```
Comment (MatousJobanek) [inline go.mod:38]: reminder

**Review by MatousJobanek — APPROVED** [review]
Cool :+1:

File: go.mod (hunk 6 above)
Comment (xcoulon) [reply go.mod]: done in 4a2a106

**Pull request merged**